fs/mmap/fs_mmap.c

```diff
@@ -142,6 +148,30 @@ FAR void *mmap(FAR void *start, size_t length, int prot, int flags,
     }
 #endif
 
+  if (fs_getfilep(fd, &filep) < 0)
```

Comment:
```suggestion
  ret = fs_getfilep(fd, &filep);
  if (ret < 0)
    {
      ferr("ERROR: Invalid file descriptor, fd=%d\n", fd);
      goto errout;
    }
```

Reply:
`fs_getfilep` either returns **-EBADF** or **-EAGAIN**.
According to the `mmap` specification, although EAGAIN is a possible return value, it is useful for indicating that a file is currently locked and the API user should try `mmap` again later.
But, `fs_getfilep` returns **-EAGAIN** for a unrecoverable error scenario:

![image](https://user-images.githubusercontent.com/38959758/111299708-07f4ca00-862f-11eb-8765-3fe71e92c48d.png)

So, I think it does not make sense to simply make `mmap` propagate the error from `fs_getfilep`. Instead, I chose to handle the error and raise a new one that fits to `mmap` API.

Reply:
errno is very useful to identify the failure location without enabling/adding debug log. It make the debug a bit harder if we eat the return code and then return the new fixed error code.

Reply:
I agree, but at the same time we should avoid simply propagating errors upward that are not compatible to the function API.

If you agree that returning EAGAIN is an error that should not happen here and does not make sense to propagate it, we could force it to fail in a DEBUGASSERT.
This way it won't impact production code and it will help the debugging during development.
What do you think?

Reply:
> errno is very useful to identify the failure location without enabling/adding debug log.

Actually, `errno` is not for debugging purposes, instead it is a error notification method for the consumer of the function API. That is why I insist that we shall only propagate error codes that are compatible to the API. All the other errors should be handled locally, even if the chosen handling is to ignore it.

Reply:
but caller should expect that the implementation may return the additional errno which not specify in the spec.

Reply:
> I agree, but at the same time we should avoid simply propagating errors upward that are not compatible to the function API.
> 
> If you agree that returning EAGAIN is an error that should not happen here and does not make sense to propagate it, we could force it to fail in a DEBUGASSERT.
> This way it won't impact production code and it will help the debugging during development.
> What do you think?

it depends on the error category. If it's programming error, we can assert it for debugging purpose. But assert is an abuse for the runtime error.

Reply:
> but caller should expect that the implementation may return the additional errno which not specify in the spec.

I agree as long as it compatible with the "Returned values" in the documentation, which was not the case for the EAGAIN returned from the `fs_getfilep`.